pkg/upgrade/drain_test.go

```diff
@@ -0,0 +1,114 @@
+package upgrade
+
+import (
+	"context"
+	"testing"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+	"github.com/openshift/openshift-azure/pkg/log"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func TestUpgraderDrain(t *testing.T) {
+	tests := []struct {
+		name            string
+		kubeclient      *fake.Clientset
+		role            api.AgentPoolProfileRole
+		nodeName        string
+		wantErr         bool
+		expectedActions [][]string
+	}{
+		{
+			name:     "master-empty",
+			role:     api.AgentPoolProfileRoleMaster,
+			nodeName: "master-000000",
+			expectedActions: [][]string{
+				{"get", "nodes"}},
+			wantErr:    false,
+			kubeclient: fake.NewSimpleClientset(),
+		},
+		{
+			name:     "unknown-role",
+			role:     "cant-find-this",
+			nodeName: "master-000000",
+			expectedActions: [][]string{
+				{"get", "nodes"}},
+			wantErr: true,
+			kubeclient: fake.NewSimpleClientset(&corev1.Node{
+				TypeMeta: metav1.TypeMeta{
+					Kind: "node",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "master-000000",
+				},
+			}),
+		},
+		{
+			name:     "master-no-pods",
+			role:     api.AgentPoolProfileRoleMaster,
+			nodeName: "master-000000",
+			expectedActions: [][]string{
+				{"get", "nodes"},
+				{"delete", "nodes"}},
+			wantErr: false,
+			kubeclient: fake.NewSimpleClientset(&corev1.Node{
+				TypeMeta: metav1.TypeMeta{
+					Kind: "node",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "master-000000",
+				},
+			}),
+		},
+		{
+			name:     "compute-with-a-pod",
+			role:     api.AgentPoolProfileRoleCompute,
+			nodeName: "kubernetes",
+			expectedActions: [][]string{
+				{"get", "nodes"},
+				{"get", "nodes"},
+				{"update", "nodes"},
+				{"list", "pods"},
+				{"delete", "pods"},
+				{"get", "pods"},
+				{"delete", "nodes"}},
+			wantErr: false,
+			kubeclient: fake.NewSimpleClientset(&corev1.Node{
+				TypeMeta: metav1.TypeMeta{
+					Kind: "node",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "kubernetes",
+				},
+			}, &corev1.Pod{
+				TypeMeta: metav1.TypeMeta{
+					Kind: "pod",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-pod",
+				},
+			}),
+		},
+	}
+	log.New(logrus.NewEntry(logrus.New()))
```

Comment:
Why do you need this?

Reply:
this log call causes an segfault otherwise: https://github.com/openshift/openshift-azure/blob/master/pkg/upgrade/drain.go#L32